Port build system to setuptools

sbird opened the pull request:
This fixes Issue #29 and allows a build with python >= 3.12

sbird committed:
Port build to modern setuptools as numpy.distutils has been removed

sbird committed:
Remove undefined property

sbird committed:
Cleanup unused leftover depenendencies

jchelly commented:
I've tried to use this branch to install nbodykit on python 3.12. It mostly works but I think installing the classylss module fails to copy the class data files over to the install location so computing the linear power spectrum in nbodykit (for example) fails because it can't find the data files.

I got it to work by copying these lines from master back into BuildCLib.run in setup.py:
```
        # copy data files from temp to classlssy package directory
        shutil.rmtree(os.path.join(self.build_lib, 'classylss', 'data'), ignore_errors=True)
        shutil.copytree(os.path.join(self.build_temp, 'data'), os.path.join(self.build_lib, 'classylss', 'data'))
```
But I have no idea if that's a reasonable fix.

sbird committed:
Add class data files back to install directory. Thanks @jchelly

sbird commented:
Seems pretty reasonable to me! Thanks.

sbird committed:
Fixes to Makefile

sbird committed:
Do not apply old patches to CLASS

sbird committed:
Do not over-write CLASS makefile with a modified old Makefile

sbird committed:
Update class version to 3.3.1

sbird committed:
Update author list

sbird committed:
Update cclassy.pxd for new classy names. spectra has become harmonic and nonlinear has become fourier. perturb is now perturbations and transfers is transfer

sbird committed:
Do renames in binding.pyx

sbird committed:
Fix init and define a spectral distortion struct

sbird committed:
a_today is not defined in class, just return unity

sbird committed:
Fix definition

sbird committed:
long_info and inter_normal are now separate enums

sbird committed:
ln_k has moved to fourier

sbird committed:
Two more renamed functions

sbird committed:
Do not decode the error message

sbird committed:
Cython module now needs C++ to build

sbird committed:
Specify CLASSDIR explicitly as the current directory, to prevent it becoming the build directory by default

sbird committed:
Explicitly specify the location of the datafiles in the python code

sbird committed:
Remember to rename tasks

sbird committed:
Need trailing slashes in paths

sbird committed:
Rename the nl struct as fo to match upstream class

sbird committed:
Use fourier_ functions to get the matter power spectrum

sbird committed:
Use class function to get sigma8

sbird committed:
Fix pointers in pk

sbird committed:
Use the classy get_transfers

sbird committed:
No longer support a_max, fix tests by removing it

sbird committed:
Class rho_crit no longer includes curvature, so we do not need rho_tot

sbird committed:
Add github action test workflow

sbird committed:
Correct rawgit links

sbird committed:
Remove old travis yml

sbird committed:
Remove old patches

sbird committed:
Remove old build files

sbird committed:
Remove runtests.py

sbird committed:
Update MANIFEST.in

sbird committed:
Update README

sbird committed:
Fix test environment

sbird committed:
Clean up some entries in pxd

sbird committed:
Don't need to free parser when using input_from_file

sbird committed:
Remove last patch

This changes the C file name length to be 512, but the python definition
was always 256 anyway, so it doesn't really make sense.
